---
title: 'Selecting the Best Inference Model'
author:
  - richelbilderbeek
date: '2020-12-01'
slug: mcbette-selecting-the-best-inference-model
categories:
  - blog
tags:
  - Software Peer Review
  - packages
  - R
  - community
  - phylogenetics
  - Bayesian
  - model selection
package_version: 0.1.0
description: How to pick the best evolutionary model for your DNA
twitterImg: blog/2019/06/04/post-template/name-of-image.png
output:
  html_document:
    keep_md: yes
---






Imagine you are a field biologist. All around the world, you captured
multiple bird species of which you obtained a blood sample.
From the blood, you have extracted the DNA. Using DNA, one can
determine how these species are evolutionarily related. The problem is,
which model of evolution you assume for your birds?

## The problem

To illustrate the problem of picking the right model of evolution,
we start from the DNA sequences of primates. DNA consists out of a long
string of four different elements called nucleotides, 
resulting in a four letter alphabet encoding for the proteins a cell needs. 
In our case, we do not have the full DNA sequence of all primates, 
but 'only' 898 nucleoides. Here I show the DNA sequences:

```r 
dna_sequences <- ape::read.FASTA(fasta_filename)
ape::image.DNAbin(dna_sequences)
```
{{<figure src="unnamed-chunk-1-1.png" >}}

From this DNA alignment, we can use the R package babette
to estimate the evolutionary history of these species:

```r 
out <- babette::bbt_run_from_model(fasta_filename)
```

An evolutionary history can be visualized by a tree-like
structure called a phylogeny. `babette`, however, creates multiple
phylogenies of which the  

```r 
babette::plot_densitree(
  out$primates_trees[9000:10000],
  alpha = 0.01
)
```
{{<figure src="unnamed-chunk-3-1.png" alt="the estimated evolutionary history of primates" caption="the estimated evolutionary history of primates" width="300">}}
mcbette 1 is a package to work with BEAST2 2, a software platform for Bayesian evolutionary analysis from R.




## Noes
The figure paths in the resulting Markdown will already be valid so you won't need to edit the Markdown file.

## Section heading in sentence case

Citation of the primary literature[^1]. 

Citation of a website[^2]. 

Citation of an R package[^3].

### Subsection heading

We recommend the use of [Hugo shortcodes](https://gohugo.io/content-management/shortcodes/) to include images, tweets, videos, gists, etc. In the Rmd file they need to be between `<!--html_preserve-->` and `<!--/html_preserve-->`.

**Add an image** by using a Hugo shortcode. The image is saved under `/content/blog/YYYY-MM-DD-slug/name-of-image.png`.

<!--html_preserve-->
{{< figure src = "name-of-image.png" width = "400" alt = "this is the alternative text" >}}
<!--/html_preserve-->

Consult the Technical Guidelines for tips on changing image size, alignment, and for advice on alternative text.

Below is another image, a plot generated by a code chunk, so you might see how to add alternative text and other options in this case. You can also use [all options documented for the figure shortcode](https://gohugo.io/content-management/shortcodes/#figure) except for `src` that will be created automatically when knitting.

```r 
plot(1:10)
```
{{<figure src="chunkname-1.png" alt="alternative text please make it informative" caption="this is what this image shows, write it here or in the paragraph after the image as you prefer" width="300">}}

Once this file is knitted the plot above will be inserted with the correct syntax.


**Embed a tweet** by using a Hugo shortcode. 

<!--html_preserve-->
{{< tweet 1138216112808529920 >}}
<!--/html_preserve-->


**Tweak code highlighting** Below you can see how to use Hugo [Chroma code highlighting options](https://gohugo.io/content-management/syntax-highlighting/#highlight-shortcode): highlight the 1st and 4th to 6th code lines, number lines as a table starting from 199.

```r {linenos=table,hl_lines=[1,"4-6"],linenostart=199}
a <- 1
a <- 1
a <- 1
a <- 1
a <- 1
a <- 1
a <- 1
```

**Add citation or footnote** text by using the format below 

[^1]: Sciaini, M., Fritsch, M., Scherer, C., & Simpkins, C. E. (2018). NLMR and landscapetools: An integrated environment for simulating and modifying neutral landscape models in R. Methods in Ecology and Evolution, 9(11), 2240-2248. <https://doi.org/10.1111/2041-210X.13076>
[^2]: Elin Waring, Michael Quinn, Amelia McNamara, Eduardo Arino de la Rubia, Hao Zhu and Shannon Ellis (2019). skimr: Compact and Flexible Summaries of Data. R package version 1.0.7. https://CRAN.R-project.org/package=skimr
[^3]: Hugo static site generator. https://gohugo.io/
