---
title: Using Open-Access Tools (rentrez, taxize) to Find Coronaviruses, Their Genetic Sequences, and Their Hosts
author: "Liam Brierley"
date: '2020-11-10'
output:
  html_document:
    keep_md: yes
  pdf_document: default
slug: coronaviruses-and-hosts
tags:
- R
- community
- packages
- taxonomy
- taxize
- epidemiology
- biology
- bioinformatics
- rentrez
- genbank
- genetics
- sars-cov-2
- covid-19
description: Together, the packages rentrez and taxize can extract standardised data on viruses and their hosts from genetic records.
---
```{r setup, include=FALSE}
# Options to have images saved in the post folder
# And to disable symbols before output
knitr::opts_chunk$set(fig.path = "", comment = "")

# knitr hook to make images output use Hugo options
knitr::knit_hooks$set(
  plot = function(x, options) {
    hugoopts <- options$hugoopts
    paste0(
      "{{<figure src=",
      '"', x, '" ',
      if (!is.null(hugoopts)) {
        glue::glue_collapse(
          glue::glue('{names(hugoopts)}="{hugoopts}"'),
          sep = " "
        )
      },
      ">}}\n"
    )
  }
)

# knitr hook to use Hugo highlighting options
knitr::knit_hooks$set(
  source = function(x, options) {
    hlopts <- options$hlopts
    paste0(
      "```r ",
      if (!is.null(hlopts)) {
        paste0("{",
               glue::glue_collapse(
                 glue::glue('{names(hlopts)}={hlopts}'),
                 sep = ","
               ), "}"
        )
      },
      "\n", glue::glue_collapse(x, sep = "\n"), "\n```\n"
    )
  }
)
```

Emerging viruses might be on everyone's mind right now, but as an epidemiologist and disease ecologist I've always been interested in how and why pathogens move from animal hosts to humans.
The current pandemic of the disease we call COVID-19 is caused by _Severe acute respiratory syndrome (SARS) coronavirus 2_ (SARS-CoV-2), a virus that has emerged from wildlife like SARS coronavirus and Middle East respiratory syndrome (MERS) coronavirus did previously. 
Although these viruses are now widely known, there are many more coronaviruses out there in nature, many of which we know little about.

There's perhaps no better time to dive into understanding where viruses like these originate.
Genome sequence repositories represent a wealthy source of data to understand the diversity of viruses and their relationships with their hosts.
However, the sheer volume of data and lack of standardisation for some features can make it challenging to use these repositories in practice.
In this post, I’ll demonstrate the use of the [rentrez](https://docs.ropensci.org/rentrez/) and [taxizedb](https://docs.ropensci.org/taxizedb/) packages to search, retrieve and resolve information about genetic sequences of coronaviruses and their animal hosts - and ask which host species have the most coronavirus sequences been sampled from?
Here's a high-level summary of the functions we'll use.

<!--/html_preserve-->
{{< figure src = "pipeline.png" width = "750" alt = "Flow diagram showing functions entrez_search(), entrez_fetch() and extract_from_esummary() from package rentrez, and name2taxid() and classification() from taxize." caption = "Flow diagram of functions used in each package to identify hosts associated with coronavirus sequences" class = "center">}}
<!--/html_preserve-->

I’ll focus on NCBI’s GenBank[^1], an open-access repository where sequences are submitted and described by individual users. 
We can look at the exponential rise in viral genome sequences by importing the release dates of those indexed by NCBI’s Genome Reports and plotting a cumulative curve.

```{r sequence_availability, warning=FALSE, hugoopts=list(alt="Curve demonstrating growth in viral genome sequences contained in GenBank since 2000", caption="Viral genome sequence availability between 2000 and present", width=850)}
library(dplyr)
library(ggplot2)
library(magrittr) # load magrittr for use of the pipe operator (%>%)

genbank_seqs <- read.table("ftp://ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/viruses.txt", 
sep = "\t", header = TRUE, comment.char = "!", quote = "", fill = TRUE)

monthly_seqs <- genbank_seqs$Release.Date %>% 
  as.Date(format="%Y/%m/%d") %>%
  cut("month") %>%
  table() %>%
  as.data.frame() %>%
  rename(month = ".", total = Freq) %>%
  mutate(month = as.Date(month), total = cumsum(total))

ggplot(monthly_seqs, aes(month, total)) +  
  geom_line(lwd=1.4) + 
  xlab('Year') +  
  ylab("Number of virus genome sequences") +
  scale_x_date(limits = as.Date(c('2000-01-01',Sys.Date()))) +
  theme_bw()
```

If we were interested in a small set of viruses, for example if we wanted to construct a specific phylogeny, we might reasonably be able to select and filter these genome sequences manually through web interfaces. 
In fact, a powerful new interface of the [NCBI Virus](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/) front-end to GenBank was unveiled for this purpose earlier this year. 
However, if we want to investigate broader viral diversity (or use a data-hungry method like neural networks), we’d benefit from a more reproducible, automated approach. 

## Obtaining virus sequences with rentrez
### Searching sequences

[rentrez](https://docs.ropensci.org/rentrez/)[^2] is a package by [David Winter](/author/david-winter/) designed to interface with various NCBI databases (collectively known as entrez), including GenBank. 
As such, it can not only conduct search requests in a single database, but also cross-reference them (see the [vignette](https://cran.r-project.org/web/packages/rentrez/vignettes/rentrez_tutorial.html) for some nice examples of this). 
For now let’s simply try and find whether there are sequences available for _Pangolin coronavirus_ in GenBank's nucleotide sequence database (i.e., `db ="nuccore"`).
This virus recently received a lot of attention because its spike protein (which is thought to be important in determining host range) is very similar to that of SARS-CoV-2[^3]^,^[^4].
We can use `entrez_search()` to conduct a general search across all fields.

```{r find_pang}
library(rentrez)
pang_cov_ids <- entrez_search(db="nuccore", term="Pangolin coronavirus")
pang_cov_ids
```

Looks like there are 14 available genome sequences for this virus - 
note that this query doesn't return the actual sequences, but the specific entry IDs. 
If we want to directly retrieve sequences, we can use `entrez_fetch()`, referencing these Pangolin coronavirus IDs.
Here I've set `rettype` to output the sequences in FASTA format, a standard bioinformatics format readable by many other packages/software.

```{r retrieve_pang, R.options=list(max.print=10)}
pang_cov_seqs <- entrez_fetch(db="nuccore", id = pang_cov_ids$ids, rettype = "fasta_cds_na")
# trim the FASTA string to show first protein coding nucleotide sequence as an example
gsub("\n\n.*", "", pang_cov_seqs) 
```

### Understanding sequence-associated metadata

However, we want to access more information about these viruses than just their nucleotide sequence.
We might want to find out information from the entry's title, or filter out incomplete sequences, etc.
`entrez_summary()` can retrieve metadata and features of the sequence as an object of class `esummary_list`.
Then we can use `extract_from_esummary()` to pull out the metadata of interest as a vector (or data frame).
Let's find out when these Pangolin coronavirus sequences were uploaded to GenBank.

```{r summary_pang}
pang_cov_summary <- entrez_summary(db="nuccore", id = pang_cov_ids$ids)
pang_cov_summary

extract_from_esummary(pang_cov_summary, elements = c("createdate")) %>% 
  table()
```

Looks like these sequences were mostly uploaded in April and August!
Note that this upload date field does not represent the _sampling_ date (this can be found within the sub-features; see below)

### More CoVs, more problems

That example's nice and simple, but in practice, if we're accessing these data programmatically we've probably got a much bigger scale of dataset in mind.
Since they're so diverse, we could look at how many sequences there are across all coronaviruses.
This time, I'll use the ID corresponding to the family [_Coronaviridae_](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=11118) in NCBI's Taxonomy (`txid11118`) and search for this in the **'Organism'** field (`[Organism]`).
By default, this will return matches for this taxonomic ID, plus any taxonomic IDs below it - meaning we're searching for all coronavirus species (should you not wish to return matches of nested taxa, use `[Organism:noexp]` instead).

```{r find_all}
all_cov_ids <- entrez_search(db="nuccore", term="txid11118[Organism]", retmax = 70000)
all_cov_ids
```

That's a lot of coronavirus sequences! By default, `entrez_search()` returns only 20 IDs, so I increased `retmax` to make sure all of them are captured.
This is also a good forewarning that there are reasonable limits to usage of eUtils (the API we're accessing underneath) and for very large queries with `entrez_fetch()` or `entrez_summary()`, we should break them up into smaller individual queries.
I've done this by looping `entrez_summary()` across our IDs and joining the results together as a single function.

```{r summary_all}

get_metadata <- function(x){
  
  query_index <- split(seq(1,length(x)), ceiling(seq_along(seq(1,length(x)))/300))
  Seq_result <- vector("list", length(query_index))
  
  for (i in 1:length(query_index)) {
    Seq_result[[i]] <- entrez_summary(db = "nuccore",id = x[unlist(query_index[i])])
    Sys.sleep(1)
  }
  
  if(length(x) == 1){
    return(Seq_result)
  } else {
    return(Seq_result %>% 
             purrr::flatten() %>% 
             unname() %>% 
             structure(class = c("list", "esummary_list"))) # coerce to esummary list
  }
}

all_cov_summary <- get_metadata(all_cov_ids$ids)
```

By using `Sys.sleep()`, this function also delays the speed at which queries are sent to comply with the [eUtils API usage limits](https://www.ncbi.nlm.nih.gov/books/NBK25497/#_chapter2_Usage_Guidelines_and_Requiremen_).
If you're likely to conduct lots of queries, you can [set up an API key to increase your usage](/blog/2018/03/20/rentrez-paper/) and [use the web history feature to quickly revisit previous queries](https://cran.r-project.org/web/packages/rentrez/vignettes/rentrez_tutorial.html).

Now we have metadata for all coronavirus sequences, let's investigate what host animals these coronaviruses came from.
You may have noticed there wasn't an explicit "host" feature among the metadata we got in our previous esummary result for Pangolin coronavirus.
This is because host is considered a sub-feature of GenBank records and (if present) it's contained in a bar-delimited string under `subtype`.
Processing this data to a tidy format requires a couple of extra steps - I've simply split these strings and reconstructed them as a data frame, adding the accession version as a unique sequence identifier.

```{r process_summary_all}
all_accession <- extract_from_esummary(all_cov_summary, elements = c("accessionversion"))
all_subname_list <- extract_from_esummary(all_cov_summary, elements = c("subname", "subtype"), 
                                          simplify=FALSE)
all_subname_list[[3000]] # show an example of sub-features

all_subname_list <- all_subname_list %>%
  lapply(function(x) {
    subname_row <- x["subname"] %>%
      stringr::str_split("\\|") %>%
      unlist() %>%
      matrix(nrow = 1, byrow = FALSE) %>%
      as.data.frame()
    subtype_names <- x["subtype"] %>%
      stringr::str_split("\\|") %>%
      unlist()
    set_colnames(subname_row, subtype_names)
  })

all_metadata_df <-  data.frame(accession = all_accession,
                               suppressMessages(bind_rows(all_subname_list)))
str(all_metadata_df[,1:21]) # show only the first 20 sub-features; there are many fields!
```

Immediately we see that not every sub-feature is filled.
In fact most are rarely ever used.
Let's take a look at the 15 most common hosts of coronaviruses according to the host sub-features.

```{r host_all}
host_table <- all_metadata_df %>% 
  filter(!is.na(host)) %>%
  count(host) %>%
  arrange(-n)

host_table %>%
  slice(1:15)
```

As you might expect, most nucleotide sequences are sampled from humans or domestic animals, particularly chickens and pigs, as there are several non-human coronaviruses that cause significant outbreaks in these animals each year.
In fact, the only wild animals in this top 15 are bats: [_Eidolon helvum_, an African fruit bat](https://en.wikipedia.org/wiki/Straw-coloured_fruit_bat), and the rather unspecific "Microchiroptera", which we'll decipher shortly.

However, because GenBank entries are filled out by the individuals depositing the data, their metadata can use inconsistent terms.
Among these top 15 hosts, we find _pig_, _swine_, _Sus scrofa_, and _piglet_, all of which describe the same animal.
The same is true for _chicken_ and _Gallus gallus_!

## Standardising taxonomies with taxizedb

This is a problem the package [taxizedb](https://docs.ropensci.org/taxizedb/)[^5] can help solve.
This is a version of the [taxize](https://docs.ropensci.org/taxize/) package by [Scott Chamberlain](/author/scott-chamberlain/) and [Zebulun Arendsee](https://github.com/arendsee). 
Its functionality is essentially similar to taxize. 
The key difference is that rather than making web queries, it references a locally stored taxonomy database, so you're not limited by API restrictions or web connectivity.

The first time you use taxizedb, you'll need to use the corresponding function to download the taxonomy database(s) you'll use.
We'll stick with NCBI's taxonomy to keep IDs consistent.
The function `db_download_ncbi()` will set everything up, though be warned this generates a fairly hefty file at approx. 2GB!

Once we have a local taxonomy database, we're ready to start resolving using `name2taxid()`, which converts various common names and synonyms to their lowest specific taxonomic ID 
(just like the ID we used for _Coronaviridae_, and if we wanted, we could equally use these taxonomic IDs to search for nucleotide sequences of hosts!)

```{r setup_taxizedb, message = FALSE}
library(taxizedb)
src_ncbi <- src_ncbi() # load the NCBI taxonomy database into R

all_tax_match <- host_table$host %>%
  name2taxid(db = "ncbi", out_type = "summary")

all_tax_match
```

Great - our synonymous strings for chickens and pigs have been matched to the same ID!
Let's also use these taxonomic ID matches to check out what is meant by the unspecific term "Microchiroptera".
Using NCBI's taxonomy web interface, ID 30560 seems to correspond to [a sub-order grouping](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=30560) within the bat order _Chiroptera_.
However, we can just as easily examine this through taxizedb.
`classification()` will return the taxonomic classification **above** the given ID, while `downstream()` will return the taxonomic classifications **below** the given ID, branching down to a specific level - so we can see some example species considered "Microchiroptera"!

```{r inv_microchiroptera, R.options=list(max.print=80)}
classification(30560, db = "ncbi")

downstream(30560, db = "ncbi", downto = "species")
```

Of course, as host metadata is often individually typed, this approach isn't infallible – typos, plus any strings that are very colloquial or ambiguous, e.g., ["broiler"](https://www.ncbi.nlm.nih.gov/nuccore/MN967777.1), or overly detailed, e.g., ["Felis catus (cat); breed Scottish Fold; color red"](https://www.ncbi.nlm.nih.gov/nuccore/FJ943767.1) may need some manual resolution.

### Where have coronaviruses been found?

Since we now have a reproducible way of retrieving the full taxonomic classifications of each host, we can now finally tie all this together and ask which hosts have the most coronavirus sequences on GenBank.
Firstly, we'll look up the `classification()` of all hosts and restructure the output as a name column and an ID column per taxonomic level from class to species.

```{r create_all_tax_full}
all_tax_full <- all_tax_match %>%
  bind_cols(
    all_tax_match$id %>%
      classification(db = "ncbi") %>%
      lapply(function(x) {
        x %>%
          data.frame() %>%
          distinct() %>%
          filter(rank == "class" | rank == "order" | rank == "family" | 
                   rank == "genus" | rank == "species") %>%
          reshape2::melt(id = "rank") %>%
          tidyr::unite(col = rank_var, c("rank", "variable")) %>%
          tidyr::spread(key = rank_var, value = value)
      }) %>%
      bind_rows() %>%
      select(class_name, class_id, order_name, order_id, family_name, family_id, 
             genus_name, genus_id, species_name, species_id)
  )

all_tax_full
```

Next, we'll attach these host taxonomic classifications to our sequences by merging these columns in to our metadata. 
From here we can count up, say, the 15 most frequent host genera among coronavirus sequences, filtering out any sequences missing host genus information.

```{r bind_in_all_tax_full}
all_metadata_plot_df <- all_metadata_df %>% 
  # match each unique sequence to its taxonomic classification
  left_join(all_tax_full, by = c("host" = "name")) %>% 
  filter(!is.na(genus_name)) %>%
  count(order_name, genus_name) %>%
  arrange(-n) %>%
  slice(1:15) %>%
  mutate(genus_name = factor(genus_name, levels = genus_name))
```
  
Lastly, we'll visualise our top 15 genera in a simple barplot using ggplot2, colour-coding each according to which taxonomic order they belong to.

```{r plot_cov_hosts, hugoopts=list(alt="Barplot showing the most frequent animal host genera of coronaviruses are Homo, Gallus, Sus and Felis", caption="Top fifteen most common animal genera represented in coronavirus genome sequence host metadata, colour-coded by order", width=1000)}
  ggplot(all_metadata_plot_df, aes(x = genus_name, y = n, fill = order_name)) +  
  geom_bar(stat = "identity") + 
  scale_y_log10(breaks = c(1, 10, 100, 1000, 10000)) +
  xlab('Host genus') +  
  ylab("Number of coronavirus genome sequences available") +
  guides(fill = guide_legend(title="order")) +
  theme_bw()
```

While there are a tremendous number of sequences of coronaviruses sampled from humans and domestic animals, there are also quite a few genera of bats present among the most represented hosts! Again, this isn't particularly surprising, as bats have a rich coevolutionary history with coronaviruses[^6], which has been an intense focus of study.

## Conclusion

rentrez offers an automated way to search and retrieve genome sequences and metadata, while taxizedb is capable of retrieving and resolving various levels of taxonomic classification.
Using these packages in combination represents a reproducible approach to extracting a standardised dataset of coronavirus nucleotide sequences and their sample hosts.
This reproducibility is essential for those studies attempting to modelling host-virus relationships to better understand future epidemic threats in humans.
I've used the above functions to process data to a suitable standard for training machine learning models to predict coronavirus hosts based on their sequence features[^7].
But of course, neither package is limited to viruses; their functionality makes for some brilliant additions to any biologist's R toolkit!


[^1]: NCBI GenBank. https://www.ncbi.nlm.nih.gov/genbank/
[^2]: Winter, D. J. (2017). rentrez: an R package for the NCBI eUtils API. The R Journal, 9(2), 520-526. <https://doi.org/10.32614/RJ-2017-058>
[^3]: Liu, P., Jiang, J.Z., Wan, X.F., Hua, Y., Li, L., Zhou, J., Wang, X., Hou, F., Chen, J., Zou, J., & Chen, J. (2020). Are pangolins the intermediate host of the 2019 novel coronavirus (SARS-CoV-2)? PLoS Pathogens, 16(5), e1008421. <https://doi.org/10.1371/journal.ppat.1008421>
[^4]: Xiao, K., Zhai, J., Feng, Y., Zhou, N., Zhang, X., Zou, J.J., Li, N., Guo, Y., Li, X., Shen, X., Zhang, Z, Chen, R.A., Wu, Y.J., Peng, S.M., Huang, M., Xie, W.J., Cai, Q.H., Hou, F.H., Chen, W., Xiao, L., & Shen, Y. (2020). Isolation of SARS-CoV-2-related coronavirus from Malayan pangolins. Nature, 583, 286-289. <https://doi.org/10.1038/s41586-020-2313-x>
[^5]: Chamberlain, S. & Arendsee, Z. (2020). taxizedb: Tools for Working with 'Taxonomic' Databases. <https://docs.ropensci.org/taxizedb>
[^6]: Cui, J., Li, F., & Shi, Z.L. (2019). Origin and evolution of pathogenic coronaviruses. Nature Reviews Microbiology, 17, 181-192. <https://doi.org/10.1038/s41579-018-0118-9>
[^7]: Brierley, L., & Fowler, A. (2020). Predicting the animal hosts of coronaviruses from compositional biases of spike protein and whole genome sequences through machine learning. bioRxiv preprint, 2020.11.02.350439. <https://doi.org/10.1101/2020.11.02.350439>
