---
slug: "rsnps-update"
title: "rsnps 0.5.0: New ncbi_snp_query() Features"
package_version: 0.5.0
author:
  - Sina RÃ¼eger
  - Julia Gustavsen
date: 2021-02-10
tags:
  - Software Peer Review
  - packages
  - R
  - community
  - rsnps
  - dbsnp
description: "`ncbi_snp_query()` now returns all reported variant allele frequencies in dbSNP."
twitterImg: blog/2022-02-03-rsnps-update/plot-maf-2.png
twitterAlt: "Allele frequencies in dbSNP for rs562556 and various studies/populations."
output: 
  html_document:
    keep_md: true
---


```{r setup, include=FALSE}
# Options to have images saved in the post folder
# And to disable symbols before output
knitr::opts_chunk$set(fig.path = "", comment = "")

# knitr hook to make images output use Hugo options
knitr::knit_hooks$set(
  plot = function(x, options) {
    hugoopts <- options$hugoopts
    paste0(
      "{{<figure src=",
      '"', x, '" ',
      if (!is.null(hugoopts)) {
        glue::glue_collapse(
          glue::glue('{names(hugoopts)}="{hugoopts}"'),
          sep = " "
        )
      },
      ">}}\n"
    )
  }
)

# knitr hook to use Hugo highlighting options
knitr::knit_hooks$set(
  source = function(x, options) {
  hlopts <- options$hlopts
    paste0(
      "```r ",
      if (!is.null(hlopts)) {
      paste0("{",
        glue::glue_collapse(
          glue::glue('{names(hlopts)}={hlopts}'),
          sep = ","
        ), "}"
        )
      },
      "\n", glue::glue_collapse(x, sep = "\n"), "\n```\n"
    )
  }
)
```


## TL;DR

`ncbi_snp_query()` now returns all reported variant allele frequencies in dbSNP in column `maf_population` in form of a `tibble`. Previously (<= 0.4.0), it reported only the allele frequency from [gnomAD](https://gnomad.broadinstitute.org/) in column `maf` as a `data.frame`. 

## Changes

The [NEWS](https://github.com/ropensci/rsnps/blob/master/NEWS.md) file will tell you about the changes of rsnps 0.4.0 to 0.5.0, but the most relevant for users we would like to highlight in this blogpost. 

`ncbi_snp_query()` is the function that pulls data from [NCBI's](https://www.ncbi.nlm.nih.gov/) [dbSNP](https://www.ncbi.nlm.nih.gov/snp/), a database of single-nucleotide polymorphisms (SNP). This database lets a user query for a SNP of interest and returns a plethora of information, among them genomic position, associated gene, clinical significance and - relevant for this blogpost - the allele frequency. The allele frequency varies typically between different populations, sometimes just a little (e.g. [rs562556](https://www.ncbi.nlm.nih.gov/snp/rs562556#frequency_tab)), sometimes a lot (e.g. [rs11677783](https://www.ncbi.nlm.nih.gov/snp/rs11677783#frequency_tab)). This is why [dbSNP](https://www.ncbi.nlm.nih.gov/snp/) collects allele frequency estimates from different studies and populations. Also, the database is constantly updated. 

Until version 0.4.0 `ncbi_snp_query()` reported the allele frequnecy estimated from [gnomAD](https://gnomad.broadinstitute.org/). For example, for SNP `rs420358` the `ncbi_snp_query` output used to look like this:

```r
ncbi_snp_query("rs420358")

#>          query chromosome        bp class         rsid          gene
#> 2     rs420358          1  40341239   snv     rs420358              

#>       alleles ancestral_allele variation_allele      seqname
#> 2     A,C,G,T                A            C,G,T NC_000001.11

#>     assembly ref_seq minor    maf
#> 2 GRCh38.p12       A     C 0.8614
```

We have now changed two things:

1. The `maf_population` column in the `ncbi_snp_query()` output contains all reported allele frequencies, not only from one study or population (the `maf` column stays the same).
2. To do that, `ncbi_snp_query` returns now a tibble (and not a dataframe). 


## Examples

Let's have a look at the output of two SNPs mentioned above. You can see that `dat` is a) a tibble and b) that the column `maf_population` is a list.


```{r setupexample, include=FALSE}
library(ggplot2)
theme_set(theme_bw())
library(dplyr)
library(tidyr)
library(forcats)
```

```{r query}
library(rsnps)
(dat <- ncbi_snp_query(c("rs11677783", "rs562556")))
```



Here is a visual to show how much the allele frequencies vary per SNP. 

First, we **flatten** the dataframe, so that each SNP and population/study are on **one line**. 
```{r maf}
(dat_maf <- dat %>% select(query, maf_population) %>% unnest(cols = c(maf_population)))
```

Then we display it with the allele frequency on the x-axis and the study along the y-axis. 

```{r plot-maf, hugoopts=list(alt="Graphical display of the allele frequency for two genetic variants (rs11677783, rs562556).", caption="Allele frequencies in dbSNP for rs11677783, rs562556 and various studies/populations.", width=300)}

ggplot(data = dat_maf %>% filter(query == "rs11677783") %>% mutate(study = forcats::fct_reorder(study, MAF ))) + 
  geom_vline(xintercept = c(0, 0.5, 1), linetype = 3, color ="gray") +
  geom_point(aes(MAF, study)) + 
  labs(title = "Allele frequency", subtitle = "rs11677783") 

ggplot(data = dat_maf %>% filter(query == "rs562556") %>% mutate(study = forcats::fct_reorder(study, MAF ))) + 
  geom_vline(xintercept = c(0, 0.5, 1), linetype = 3, color ="gray") +
  geom_point(aes(MAF, study)) + 
  labs(title = "Allele frequency", subtitle = "rs562556") 

```

We can decide to turn the tibble into a dataframe again and pick a **specific study** (note that `maf_korean` is a `dbl` again): 

```{r mafkorean}
(dat_korean <- 
  dat %>% mutate(maf_korean = purrr::map(maf_population, ~..1$MAF[..1$study=="KOREAN"])) %>% unnest(cols = c(maf_korean)))
                  
```

Lastly, we can decide to pivot the study allele frequencies, so that **each study** has its **own column**:

```{r mafcolumn}
       
dat_maf <- dat %>% select(query, maf_population) %>% unnest(cols = c(maf_population)) %>%
  select(query, study, MAF) %>%
  pivot_wider(values_from = "MAF", names_from = "study", values_fn = min, names_prefix = "maf_") ## if duplicate, picking the minimum

(dat_wide <- 
  dat %>% left_join(dat_maf, by = "query"))
           
```